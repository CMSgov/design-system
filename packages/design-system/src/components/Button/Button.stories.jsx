import React from 'react';

import Button from './Button';
import { NextIcon } from '../Icons';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    // children can be text or a ReactNode. For storybook controls, just allowing strings
    children: {
      control: { type: 'text' },
      type: { name: 'string', required: true },
    },
    component: {
      control: false,
    },
    disabled: {
      control: { type: 'boolean' },
    },
    // hiding deprecated props
    inverse: {
      table: {
        disable: true,
      },
    },
    // hiding deprecated opts
    variation: {
      options: ['primary', 'success', 'transparent'],
    },
  },
  args: {
    children: 'Your button text here',
  },
  parameters: {
    backgrounds: {
      default: 'light',
      values: [
        {
          name: 'light',
          value: '#fff',
        },
        {
          name: 'Hcgov dark',
          value: '#112e51',
        },
        {
          name: 'Mgov dark',
          value: '#146a5d',
        },
      ],
    },
  },
};

const Template = ({ data, ...args }) => <Button {...args} />;

export const DefaultButton = Template.bind({});
export const InverseButton = Template.bind({});
InverseButton.args = {
  inversed: true,
};
InverseButton.parameters = {
  backgrounds: { default: process.env.STORYBOOK_DS === 'mgov' ? 'Mgov dark' : 'Hcgov dark' },
};

export const IconButton = Template.bind({});
IconButton.args = {
  children: (
    <>
      Button with icon <NextIcon />{' '}
    </>
  ),
};
