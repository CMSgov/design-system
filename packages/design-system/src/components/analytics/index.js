export { default as sendEvent } from './Analytics';
