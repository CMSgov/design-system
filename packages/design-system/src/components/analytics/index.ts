export { default as sendAnalytics } from './SendAnalytics';
