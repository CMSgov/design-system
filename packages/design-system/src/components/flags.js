// featureFlags.js
const flags = {
  ERROR_PLACEMENT_DEFAULT: 'top',
};

export function errorPlacementDefault() {
  return flags.ERROR_PLACEMENT_DEFAULT;
}

export function setErrorPlacementDefault(value) {
  flags.ERROR_PLACEMENT_DEFAULT = value;
}
