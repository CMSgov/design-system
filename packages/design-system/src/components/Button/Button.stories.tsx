import React from 'react';
import Button from './Button';
import { NextIcon } from '../Icons';
import Spinner from '../Spinner/Spinner';
import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  args: {
    children: 'Your button text here',
    onDark: false,
  },
};
export default meta;

type Story = StoryObj<typeof Button>;

export const Default: Story = {};

export const OnDark: Story = {
  args: {
    onDark: true,
  },
  parameters: {
    layout: 'fullscreen',
    onDark: true,
  },
};

export const IconButton: Story = {
  args: {
    children: (
      <>
        Button with icon <NextIcon />{' '}
      </>
    ),
  },
};

export const SpinnerButton: Story = {
  args: {
    children: (
      <>
        <Spinner /> Loading...
      </>
    ),
    variation: 'solid',
  },
};

export const AnchorButton: Story = {
  args: {
    href: '!#',
  },
};

export const VariationsOnDark: Story = {
  decorators: [
    (Story) => (
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    // Must supply `layout: 'fullscreen'` when we use `onDark: true`
    onDark: true,
    layout: 'fullscreen',
  },
  render: () => (
    <>
      <Button onDark>Outline</Button>
      <Button onDark variation="solid">
        Solid
      </Button>
      <Button onDark variation="ghost">
        Ghost
      </Button>
    </>
  ),
};

export const AllAnchorButtons: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }} className="ds-content">
      <h4>Default and modified default</h4>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }} className="">
        <Button href="#!" size="big">
          Anchor button
        </Button>
        <Button href="#!">Anchor button</Button>
        <Button href="#!" size="small">
          Anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" size="big" disabled>
          Anchor button
        </Button>
        <Button href="#!" disabled>
          Anchor button
        </Button>
        <Button href="#!" size="small" disabled>
          Anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" size="big" isAlternate>
          Alt anchor button
        </Button>
        <Button href="#!" isAlternate>
          Alt anchor button
        </Button>
        <Button href="#!" size="small" isAlternate>
          Alt anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" size="big" disabled isAlternate>
          Alt anchor button
        </Button>
        <Button href="#!" disabled isAlternate>
          Alt anchor button
        </Button>
        <Button href="#!" size="small" disabled isAlternate>
          Alt anchor button
        </Button>
      </div>
      <div
        className="ds-u-fill--base ds-u-padding--2"
        style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }}
      >
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark size="big">
            Anchor button on-dark
          </Button>
          <Button href="#!" onDark>
            Anchor button on-dark
          </Button>
          <Button href="#!" onDark size="small">
            Anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark size="big" disabled>
            Anchor button on-dark
          </Button>
          <Button href="#!" onDark disabled>
            Anchor button on-dark
          </Button>
          <Button href="#!" onDark size="small" disabled>
            Anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark size="big" isAlternate>
            Alt anchor button on-dark
          </Button>
          <Button href="#!" onDark isAlternate>
            Alt anchor button on-dark
          </Button>
          <Button href="#!" onDark size="small" isAlternate>
            Alt anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark size="big" disabled isAlternate>
            Alt anchor button on-dark
          </Button>
          <Button href="#!" onDark disabled isAlternate>
            Alt anchor button on-dark
          </Button>
          <Button href="#!" onDark size="small" disabled isAlternate>
            Alt anchor button on-dark
          </Button>
        </div>
      </div>

      <h4>Solids</h4>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="solid" size="big">
          Solid anchor button
        </Button>
        <Button href="#!" variation="solid">
          Solid anchor button
        </Button>
        <Button href="#!" variation="solid" size="small">
          Solid anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="solid" size="big" disabled>
          Solid anchor button
        </Button>
        <Button href="#!" variation="solid" disabled>
          Solid anchor button
        </Button>
        <Button href="#!" variation="solid" size="small" disabled>
          Solid anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="solid" size="big" isAlternate>
          Solid alt anchor button
        </Button>
        <Button href="#!" variation="solid" isAlternate>
          Solid alt anchor button
        </Button>
        <Button href="#!" variation="solid" size="small" isAlternate>
          Solid alt anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="solid" size="big" disabled isAlternate>
          Solid alt anchor button
        </Button>
        <Button href="#!" variation="solid" disabled isAlternate>
          Solid alt anchor button
        </Button>
        <Button href="#!" variation="solid" size="small" disabled isAlternate>
          Solid alt anchor button
        </Button>
      </div>
      <div
        className="ds-u-fill--base ds-u-padding--2"
        style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }}
      >
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="solid" size="big">
            Solid anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid">
            Solid anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid" size="small">
            Solid anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="solid" size="big" disabled>
            Solid anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid" disabled>
            Solid anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid" size="small" disabled>
            Solid anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="solid" size="big" isAlternate>
            Solid alt anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid" isAlternate>
            Solid alt anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid" size="small" isAlternate>
            Solid alt anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="solid" size="big" disabled isAlternate>
            Solid alt anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid" disabled isAlternate>
            Solid anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="solid" size="small" disabled isAlternate>
            Solid alt anchor button on-dark
          </Button>
        </div>
      </div>

      <h4>Ghosts</h4>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="ghost" size="big">
          Ghost anchor button
        </Button>
        <Button href="#!" variation="ghost">
          Ghost anchor button
        </Button>
        <Button href="#!" variation="ghost" size="small">
          Ghost anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="ghost" size="big" disabled>
          Ghost anchor button
        </Button>
        <Button href="#!" variation="ghost" disabled>
          Ghost anchor button
        </Button>
        <Button href="#!" variation="ghost" size="small" disabled>
          Ghost anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="ghost" size="big" isAlternate>
          Ghost alt anchor button
        </Button>
        <Button href="#!" variation="ghost" isAlternate>
          Ghost alt anchor button
        </Button>
        <Button href="#!" variation="ghost" size="small" isAlternate>
          Ghost alt anchor button
        </Button>
      </div>
      <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button href="#!" variation="ghost" size="big" disabled isAlternate>
          Ghost alt anchor button
        </Button>
        <Button href="#!" variation="ghost" disabled isAlternate>
          Ghost alt anchor button
        </Button>
        <Button href="#!" variation="ghost" size="small" disabled isAlternate>
          Ghost alt anchor button
        </Button>
      </div>
      <div
        className="ds-u-fill--base ds-u-padding--2"
        style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }}
      >
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="ghost" size="big">
            Ghost anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost">
            Ghost anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost" size="small">
            Ghost anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="ghost" size="big" disabled>
            Ghost anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost" disabled>
            Ghost anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost" size="small" disabled>
            Ghost anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="ghost" size="big" isAlternate>
            Ghost alt anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost" isAlternate>
            Ghost alt anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost" size="small" isAlternate>
            Ghost alt anchor button on-dark
          </Button>
        </div>
        <div className="" style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button href="#!" onDark variation="ghost" size="big" disabled isAlternate>
            Ghost alt anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost" disabled isAlternate>
            Ghost alt anchor button on-dark
          </Button>
          <Button href="#!" onDark variation="ghost" size="small" disabled isAlternate>
            Ghost alt anchor button on-dark
          </Button>
        </div>
      </div>
    </div>
  ),
};

export const AllButtons: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }} className="ds-content">
      <h4>Default and modified default</h4>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button size="big">Button</Button>
        <Button>Button</Button>
        <Button size="small">Button</Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button size="big" disabled>
          Button
        </Button>
        <Button disabled>Button</Button>
        <Button size="small" disabled>
          Button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button size="big" isAlternate>
          Alt button
        </Button>
        <Button isAlternate>Alt button</Button>
        <Button size="small" isAlternate>
          Alt button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button size="big" disabled isAlternate>
          Alt button
        </Button>
        <Button disabled isAlternate>
          Alt button
        </Button>
        <Button size="small" disabled isAlternate>
          Alt button
        </Button>
      </div>
      <div
        className="ds-u-fill--base ds-u-padding--2"
        style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }}
      >
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark size="big">
            Button on-dark
          </Button>
          <Button onDark>Button on-dark</Button>
          <Button onDark size="small">
            Button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark size="big" disabled>
            Button on-dark
          </Button>
          <Button onDark disabled>
            Button on-dark
          </Button>
          <Button onDark size="small" disabled>
            Button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark size="big" isAlternate>
            Alt button on-dark
          </Button>
          <Button onDark isAlternate>
            Alt button on-dark
          </Button>
          <Button onDark size="small" isAlternate>
            Alt button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark size="big" disabled isAlternate>
            Alt button on-dark
          </Button>
          <Button onDark disabled isAlternate>
            Alt button on-dark
          </Button>
          <Button onDark size="small" disabled isAlternate>
            Alt button on-dark
          </Button>
        </div>
      </div>

      <h4>Solids</h4>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="solid" size="big">
          Solid button
        </Button>
        <Button variation="solid">Solid button</Button>
        <Button variation="solid" size="small">
          Solid button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="solid" size="big" disabled>
          Solid button
        </Button>
        <Button variation="solid" disabled>
          Solid button
        </Button>
        <Button variation="solid" size="small" disabled>
          Solid button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="solid" size="big" isAlternate>
          Solid alt button
        </Button>
        <Button variation="solid" isAlternate>
          Solid alt button
        </Button>
        <Button variation="solid" size="small" isAlternate>
          Solid alt button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="solid" size="big" disabled isAlternate>
          Solid alt button
        </Button>
        <Button variation="solid" disabled isAlternate>
          Solid alt button
        </Button>
        <Button variation="solid" size="small" disabled isAlternate>
          Solid alt button
        </Button>
      </div>
      <div
        className="ds-u-fill--base ds-u-padding--2"
        style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }}
      >
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="solid" size="big">
            Solid button on-dark
          </Button>
          <Button onDark variation="solid">
            Solid button on-dark
          </Button>
          <Button onDark variation="solid" size="small">
            Solid button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="solid" size="big" disabled>
            Solid button on-dark
          </Button>
          <Button onDark variation="solid" disabled>
            Solid button on-dark
          </Button>
          <Button onDark variation="solid" size="small" disabled>
            Solid button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="solid" size="big" isAlternate>
            Solid alt button on-dark
          </Button>
          <Button onDark variation="solid" isAlternate>
            Solid alt button on-dark
          </Button>
          <Button onDark variation="solid" size="small" isAlternate>
            Solid alt button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="solid" size="big" disabled isAlternate>
            Solid alt button on-dark
          </Button>
          <Button onDark variation="solid" disabled isAlternate>
            Solid button on-dark
          </Button>
          <Button onDark variation="solid" size="small" disabled isAlternate>
            Solid alt button on-dark
          </Button>
        </div>
      </div>

      <h4>Ghosts</h4>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="ghost" size="big">
          Ghost button
        </Button>
        <Button variation="ghost">Ghost button</Button>
        <Button variation="ghost" size="small">
          Ghost button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="ghost" size="big" disabled>
          Ghost button
        </Button>
        <Button variation="ghost" disabled>
          Ghost button
        </Button>
        <Button variation="ghost" size="small" disabled>
          Ghost button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="ghost" size="big" isAlternate>
          Ghost alt button
        </Button>
        <Button variation="ghost" isAlternate>
          Ghost alt button
        </Button>
        <Button variation="ghost" size="small" isAlternate>
          Ghost alt button
        </Button>
      </div>
      <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
        <Button variation="ghost" size="big" disabled isAlternate>
          Ghost alt button
        </Button>
        <Button variation="ghost" disabled isAlternate>
          Ghost alt button
        </Button>
        <Button variation="ghost" size="small" disabled isAlternate>
          Ghost alt button
        </Button>
      </div>
      <div
        className="ds-u-fill--base ds-u-padding--2"
        style={{ display: 'flex', gap: '1rem', flexDirection: 'column' }}
      >
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="ghost" size="big">
            Ghost button on-dark
          </Button>
          <Button onDark variation="ghost">
            Ghost button on-dark
          </Button>
          <Button onDark variation="ghost" size="small">
            Ghost button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="ghost" size="big" disabled>
            Ghost button on-dark
          </Button>
          <Button onDark variation="ghost" disabled>
            Ghost button on-dark
          </Button>
          <Button onDark variation="ghost" size="small" disabled>
            Ghost button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="ghost" size="big" isAlternate>
            Ghost alt button on-dark
          </Button>
          <Button onDark variation="ghost" isAlternate>
            Ghost alt button on-dark
          </Button>
          <Button onDark variation="ghost" size="small" isAlternate>
            Ghost alt button on-dark
          </Button>
        </div>
        <div style={{ display: 'flex', gap: '1rem', alignItems: 'center' }}>
          <Button onDark variation="ghost" size="big" disabled isAlternate>
            Ghost alt button on-dark
          </Button>
          <Button onDark variation="ghost" disabled isAlternate>
            Ghost alt button on-dark
          </Button>
          <Button onDark variation="ghost" size="small" disabled isAlternate>
            Ghost alt button on-dark
          </Button>
        </div>
      </div>
    </div>
  ),
};
