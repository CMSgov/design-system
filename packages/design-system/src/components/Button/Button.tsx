import React from 'react';
import classNames from 'classnames';
import mergeRefs from '../utilities/mergeRefs';
import useButtonAnalytics from './useButtonAnalytics';
import { AnalyticsOverrideProps, AnalyticsParentDataProps } from '../analytics';
import { MutableRefObject } from 'react';

export type ButtonSize = 'small' | 'big';

export type ButtonVariation = 'solid' | 'ghost';

export type ButtonRef = MutableRefObject<any> | ((...args: any[]) => any);

interface CommonButtonProps extends AnalyticsOverrideProps, AnalyticsParentDataProps {
  /**
   * Label text or HTML
   */
  children: string | React.ReactNode;
  /**
   * Additional classes to be added to the root button element.
   */
  className?: string;
  disabled?: boolean;
  /**
   * When provided, the root component will render as an `<a>` element
   * rather than `button`.
   */
  href?: string;
  /**
   * Access a reference to the `button` or `a` element
   */
  inputRef?: ButtonRef;
  /**
   * Applies the alternate color to a Button. By default, Button
   * uses the `main` color.
   */
  isAlternate?: boolean;
  /**
   * Returns the [`SyntheticEvent`](https://facebook.github.io/react/docs/events.html).
   * Not called when the Button is disabled.
   */
  onClick?: (...args: any[]) => any;
  /**
   * Defines a color palette best used when Button is placed on a dark
   * background-color. By default, Button uses a light color palette.
   */
  onDark?: boolean;
  size?: ButtonSize;
  /**
   * Button [`type`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type) attribute
   */
  type?: React.ComponentProps<'button'>['type'];
  /**
   * A string corresponding to Button variation classes.
   */
  variation?: ButtonVariation;
}

// Collect all the additional properties that one could supply to a button component
// that will be passed down to whatever element or component is being used. This is
// generally permissive in order to keep the typing simple at the expense of being
// more accurate. In a previous version of this, `OtherProps` was generic so that we
// could extract any props from a custom component that is being passed in; however,
// we are deprecating that prop because it's not actually needed and creates
// unnecessary complexity that we have to maintain.
type OtherProps = Omit<
  // All other props that could be passed to buttons or anchors
  React.ComponentPropsWithRef<'button'> & React.ComponentPropsWithRef<'a'>,
  // Omit any properties that we're defining on our own `CommonButtonProps`
  keyof CommonButtonProps
>;

export type ButtonProps = CommonButtonProps & OtherProps;

/**
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/button/).
 */
export const Button = (props: ButtonProps) => {
  const {
    analytics,
    analyticsLabelOverride,
    analyticsEventTypeOverride,
    analyticsParentHeading,
    analyticsParentType,
    onAnalyticsEvent,
    children,
    className,
    inputRef,
    isAlternate,
    onClick,
    onDark,
    size,
    variation,
    ...otherProps
  } = props;

  const { contentRef, sendButtonEvent } = useButtonAnalytics(props);
  const ComponentType = props.href ? 'a' : 'button';
  const colorSchemeClass = isAlternate && `ds-c-button--alternate`;
  const modeClass = onDark && `ds-c-button--on-dark`;
  const sizeClass = size && `ds-c-button--${size}`;
  const variationClass = variation && `ds-c-button--${variation}`;

  const allClassNames = classNames(
    'ds-c-button',
    colorSchemeClass,
    modeClass,
    sizeClass,
    variationClass,
    className
  );

  const attrs = {
    ...otherProps,
    className: allClassNames,
  };

  if (ComponentType !== 'button') {
    delete attrs.disabled;
    delete attrs.type;

    if (props.disabled) {
      attrs.role = 'link';
      attrs['aria-disabled'] = true;
      delete attrs.href;
    }
  }

  function handleClick(e: React.MouseEvent | React.KeyboardEvent): void {
    if (!props.disabled) {
      sendButtonEvent();
      if (onClick) {
        onClick(e);
      }
    }
  }

  function handleKeyPress(e: React.KeyboardEvent): void {
    // Trigger onClick on space key event for `<a>` elements
    if (e.key === ' ') {
      handleClick(e);
    }
  }

  return (
    <ComponentType
      ref={mergeRefs([inputRef, contentRef])}
      onClick={handleClick}
      onKeyPress={ComponentType === 'a' ? handleKeyPress : undefined}
      {...attrs}
    >
      {children}
    </ComponentType>
  );
};

Button.defaultProps = {
  isAlternate: false,
  onDark: false,
  type: 'button' as const,
};

export default Button;
