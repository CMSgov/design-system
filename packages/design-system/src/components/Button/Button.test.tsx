import React from 'react';
import Button from './Button';
import { UtagContainer } from '../analytics';
import { setButtonSendsAnalytics } from '../flags';
import { fireEvent, render, screen } from '@testing-library/react';

const defaultProps = {
  children: 'Foo',
};

function renderButton(props = {}) {
  return render(<Button {...defaultProps} {...props} />);
}

describe('Button', () => {
  it('renders as button', () => {
    renderButton();
    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('renders as submit button', () => {
    renderButton({ type: 'submit' });
    expect(screen.getByRole('button').getAttribute('type')).toEqual('submit');
  });

  it('renders disabled button', () => {
    renderButton({ disabled: true });
    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('renders as an anchor with custom prop', () => {
    renderButton({
      href: '/example',
      target: '_blank',
      type: 'submit',
    });
    expect(screen.getByRole('link')).toMatchSnapshot();
  });

  it('renders disabled anchor correctly', () => {
    renderButton({
      href: '#!',
      disabled: true,
      children: 'Link button',
    });
    expect(screen.getByRole('link')).toMatchSnapshot();
  });

  it('applies additional classes', () => {
    renderButton({ className: 'foobar' });
    const button = screen.getByRole('button');
    expect(button.classList.contains('foobar')).toBe(true);
    expect(button.classList.contains('ds-c-button')).toBe(true);
  });

  it('applies variation classes', () => {
    renderButton({ variation: 'solid' });
    const button = screen.getByRole('button');
    expect(button.classList.contains('ds-c-button')).toBe(true);
    expect(button.classList.contains('ds-c-button--solid')).toBe(true);
  });

  it('applies size classes', () => {
    renderButton({ size: 'small' });
    const button = screen.getByRole('button');
    expect(button.classList.contains('ds-c-button')).toBe(true);
    expect(button.classList.contains('ds-c-button--small')).toBe(true);
  });

  it('applies disabled, inverse, alternate, and variation classes together', () => {
    renderButton({
      href: '#!',
      disabled: true,
      onDark: true,
      isAlternate: true,
      variation: 'ghost',
    });
    const link = screen.getByRole('link');
    expect(link.hasAttribute('href')).toBe(false);
    expect(link.classList.contains('ds-c-button--ghost')).toBe(true);
    expect(link.classList.contains('ds-c-button--on-dark')).toBe(true);
    expect(link.classList.contains('ds-c-button--alternate')).toBe(true);
    expect(link.classList.contains('ds-c-button')).toBe(true);
  });

  describe('Analytics', () => {
    let tealiumMock;

    beforeEach(() => {
      setButtonSendsAnalytics(true);
      tealiumMock = jest.fn();
      (window as any as UtagContainer).utag = {
        link: tealiumMock,
      };
    });

    afterEach(() => {
      setButtonSendsAnalytics(false);
      jest.resetAllMocks();
    });

    it('sends button analytics event', () => {
      renderButton();
      fireEvent.click(screen.getByRole('button'));
      expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
    });

    it('sends anchor analytics event', () => {
      renderButton({ href: '#/somewhere-over-the-rainbow' });
      fireEvent.click(screen.getByRole('link'));
      expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
    });

    it('disables analytics event tracking', () => {
      renderButton({ analytics: false });
      fireEvent.click(screen.getByRole('button'));
      expect(tealiumMock).not.toBeCalled();
    });

    it('setting analytics to true overrides flag value', () => {
      setButtonSendsAnalytics(false);
      renderButton({ analytics: true });
      fireEvent.click(screen.getByRole('button'));
      expect(tealiumMock).toHaveBeenCalled();
    });

    it('overrides analytics event tracking on open', () => {
      renderButton({ analyticsLabelOverride: 'alternate content' });
      fireEvent.click(screen.getByRole('button'));
      expect(tealiumMock.mock.calls[0]).toMatchSnapshot();
    });

    it('passes along parent heading and type', () => {
      const analyticsParentHeading = 'Hello World';
      const analyticsParentType = 'div';
      renderButton({
        analyticsParentHeading,
        analyticsParentType,
      });
      fireEvent.click(screen.getByRole('button'));
      expect(tealiumMock).toBeCalledWith(
        expect.objectContaining({
          parent_component_heading: analyticsParentHeading,
          parent_component_type: analyticsParentType,
        })
      );
    });
  });
});
