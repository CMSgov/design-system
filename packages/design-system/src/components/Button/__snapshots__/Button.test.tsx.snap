// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Analytics sends button analytics event 1`] = `
Array [
  Object {
    "button_style": "default",
    "button_type": "button",
    "event_name": "button_engagement",
    "event_type": "ui interaction",
    "ga_eventAction": "engaged default button",
    "ga_eventCategory": "ui interaction",
    "ga_eventLabel": "Foo",
    "ga_eventType": "cmsds",
    "ga_eventValue": "",
    "parent_component_heading": " ",
    "parent_component_type": " ",
    "text": "Foo",
  },
]
`;

exports[`Button Analytics sends link analytics event 1`] = `
Array [
  Object {
    "button_style": "default",
    "button_type": "link",
    "event_name": "button_engagement",
    "event_type": "ui interaction",
    "ga_eventAction": "engaged default button",
    "ga_eventCategory": "ui interaction",
    "ga_eventLabel": "Foo: #/somewhere-over-the-rainbow",
    "ga_eventType": "cmsds",
    "ga_eventValue": "",
    "link_url": "#/somewhere-over-the-rainbow",
    "parent_component_heading": " ",
    "parent_component_type": " ",
    "text": "Foo",
  },
]
`;

exports[`Button renders as a custom Link component 1`] = `
<a
  class="ds-c-button"
  href="#"
  to="anywhere"
>
  Foo
</a>
`;

exports[`Button renders as an anchor with custom prop 1`] = `
<a
  class="ds-c-button"
  href="/example"
  target="_blank"
>
  Foo
</a>
`;

exports[`Button renders as button 1`] = `
<button
  class="ds-c-button"
  type="button"
>
  Foo
</button>
`;

exports[`Button renders disabled button 1`] = `
<button
  class="ds-c-button"
  disabled=""
  type="button"
>
  Foo
</button>
`;

exports[`Button renders disabled link correctly 1`] = `
<a
  class="ds-c-button ds-c-button--disabled"
  href="javascript:void(0)"
>
  Link button
</a>
`;
